tests/test_config_entries.py

```diff
@@ -1577,3 +1577,198 @@ async def async_step_import(self, user_input):
+        )
+        assert result["type"] == data_entry_flow.RESULT_TYPE_ABORT
+        assert result["reason"] == "already_configured"
+
+
+async def test_default_discovery_in_progress(hass, manager):
+    """Test that a flow using default discovery can only be triggered once."""
+    mock_integration(hass, MockModule("comp"))
+    mock_entity_platform(hass, "config_flow.comp", None)
+
+    class TestFlow(config_entries.ConfigFlow):
+        """Test flow."""
+
+        VERSION = 1
+
+        async def async_step_discovery(self, user_input=None):
+            """Test discovery step."""
+            await self.async_set_unique_id(user_input.get("unique_id"))
+            await self.async_handle_discovery_without_unique_id()
+            return self.async_show_form(step_id="mock")
+
+    with patch.dict(config_entries.HANDLERS, {"comp": TestFlow}):
+        result = await manager.flow.async_init(
+            "comp",
+            context={"source": config_entries.SOURCE_DISCOVERY},
+            data={"unique_id": "mock-unique-id"},
+        )
+        assert result["type"] == data_entry_flow.RESULT_TYPE_FORM
+
+        # Second discovery without a unique ID
+        result2 = await manager.flow.async_init(
+            "comp", context={"source": config_entries.SOURCE_DISCOVERY}, data={}
+        )
+        assert result2["type"] == data_entry_flow.RESULT_TYPE_ABORT
+
+    flows = hass.config_entries.flow.async_progress()
+    assert len(flows) == 1
+    assert flows[0]["context"]["unique_id"] == "mock-unique-id"
+
+
+async def test_default_discovery_abort_on_new_unique_flow(hass, manager):
+    """Test that a flow using default discovery is aborted when a second flow with unique ID is created."""
+    mock_integration(hass, MockModule("comp"))
+    mock_entity_platform(hass, "config_flow.comp", None)
+
+    class TestFlow(config_entries.ConfigFlow):
+        """Test flow."""
+
+        VERSION = 1
+
+        async def async_step_discovery(self, user_input=None):
+            """Test discovery step."""
+            await self.async_set_unique_id(user_input.get("unique_id"))
+            await self.async_handle_discovery_without_unique_id()
+            return self.async_show_form(step_id="mock")
+
+    with patch.dict(config_entries.HANDLERS, {"comp": TestFlow}):
+        # First discovery with default, no unique ID
+        result2 = await manager.flow.async_init(
+            "comp", context={"source": config_entries.SOURCE_DISCOVERY}, data={}
+        )
+        assert result2["type"] == data_entry_flow.RESULT_TYPE_FORM
+
+        # Second discovery brings in a unique ID
+        result = await manager.flow.async_init(
+            "comp",
+            context={"source": config_entries.SOURCE_DISCOVERY},
+            data={"unique_id": "mock-unique-id"},
+        )
+        assert result["type"] == data_entry_flow.RESULT_TYPE_FORM
+
+    # Ensure the first one is cancelled and we end up with just the last one
```

Comment:
I'm confused by this logic. When can this happen? If this can happen when both SSDP and Zeroconf are configured, would we then require the config flow to call `async_handle_discovery_without_unique_id` ? 

I think that we should just expect developers that implement discovery to implement all types of discovery.

Reply:
See comments above :)